### PR TITLE
Issue #65  ValidateTwoFactorPin always returns false, if the secretKey parameter is base32 encoded string - fixed

### PR DESCRIPTION
**Pull request by caterina-novak**
Fix for issue # 65;
Support for .net5;
Small refactoring.

### CODE REVIEW

File: Google.Authenticator/Base32Encoding.cs [hunk 1]
```diff
@@ -15,20 +15,18 @@ public class Base32Encoding
         public static byte[] ToBytes(string input)
         {
             if (string.IsNullOrEmpty(input))
-            {
-                throw new ArgumentNullException("input");
-            }
+                throw new ArgumentNullException(nameof(input));
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:18]: Please don't remove {} here :)

File: Google.Authenticator/Base32Encoding.cs [hunk 2]
```diff
@@ -48,9 +46,7 @@ public static byte[] ToBytes(string input)
 
             //if we didn't end with a full byte
             if (arrayIndex != byteCount)
-            {
                 returnArray[arrayIndex] = curByte;
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:49]: Please don't remove {} here

File: Google.Authenticator/Base32Encoding.cs [hunk 3]
```diff
@@ -63,17 +59,15 @@ public static byte[] ToBytes(string input)
         public static string ToString(byte[] input)
         {
             if (input == null || input.Length == 0)
-            {
-                throw new ArgumentNullException("input");
-            }
+                throw new ArgumentNullException(nameof(input));
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:62]: Please don't remove {} here

File: Google.Authenticator/Base32Encoding.cs [hunk 4]
```diff
@@ -101,40 +95,32 @@ public static string ToString(byte[] input)
 
         private static int CharToValue(char c)
         {
-            int value = (int)c;
+            var value = (int)c;
 
             //65-90 == uppercase letters
             if (value < 91 && value > 64)
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:101]: Please don't remove {} here

File: Google.Authenticator/Base32Encoding.cs [hunk 5]
```diff
@@ -101,40 +95,32 @@ public static string ToString(byte[] input)
...
+            var value = (int)c;
 
             //65-90 == uppercase letters
             if (value < 91 && value > 64)
-            {
                 return value - 65;
-            }
+            
             //50-55 == numbers 2-7
             if (value < 56 && value > 49)
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:105]: Please don't remove {} here

File: Google.Authenticator/Base32Encoding.cs [hunk 6]
```diff
@@ -101,40 +95,32 @@ public static string ToString(byte[] input)
...
-            {
                 return value - 24;
-            }
+           
             //97-122 == lowercase letters
             if (value < 123 && value > 96)
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:109]: Please don't remove {} here

File: Google.Authenticator/Base32Encoding.cs [hunk 7]
```diff
@@ -101,40 +95,32 @@ public static string ToString(byte[] input)
...
                 return value - 97;
-            }
 
-            throw new ArgumentException("Character is not a Base32 character.", "c");
+            throw new ArgumentException("Character is not a Base32 character.", nameof(c));
         }
 
         private static char ValueToChar(byte b)
         {
             if (b < 26)
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:117]: Please don't remove {} here

File: Google.Authenticator/Base32Encoding.cs [hunk 8]
```diff
@@ -101,40 +95,32 @@ public static string ToString(byte[] input)
...
-            {
                 return (char)(b + 65);
-            }
 
             if (b < 32)
```
Comment (flytzen) [inline Google.Authenticator/Base32Encoding.cs:120]: Please don't remove {} here

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 9]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
-        {
-            if (String.IsNullOrWhiteSpace(accountTitleNoSpaces)) { throw new NullReferenceException("Account Title is null"); }
+        public SetupCode GenerateSetupCode(
+            string issuer,
+            string accountTitleNoSpaces,
+            byte[] accountSecretKey,
+            int qrPixelsPerModule = 3,
+            bool generateQrCode = true)
+        {
+            if (string.IsNullOrWhiteSpace(accountTitleNoSpaces))
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:69]: Please don't remove {} here

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 10]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
-                //  https://github.com/google/google-authenticator/wiki/Conflicting-Accounts
-                // Added additional prefix to account otpauth://totp/Company:joe_example@gmail.com for backwards compatibility
-                provisionUrl = String.Format("otpauth://totp/{2}:{0}?secret={1}&issuer={2}", accountTitleNoSpaces, encodedSecretKey.Trim('='), UrlEncode(issuer));
-            }
+            var encodedSecretKey = Base32Encoding.ToString(accountSecretKey);
+            var provisionUrl = string.IsNullOrWhiteSpace(issuer)
+                ? $"otpauth://totp/{accountTitleNoSpaces}?secret={encodedSecretKey.Trim('=')}"
+                : $"otpauth://totp/{UrlEncode(issuer)}:{accountTitleNoSpaces}?secret={encodedSecretKey.Trim('=')}&issuer={UrlEncode(issuer)}";
+
+            if (!generateQrCode)
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:78]: Please don't remove {} here

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 11]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
             accountTitleNoSpaces = RemoveWhitespace(Uri.EscapeUriString(accountTitleNoSpaces));
-            string encodedSecretKey = Base32Encoding.ToString(accountSecretKey);
-            string provisionUrl;
-            if (String.IsNullOrWhiteSpace(issuer))
-            {
-                provisionUrl = String.Format("otpauth://totp/{0}?secret={1}", accountTitleNoSpaces, encodedSecretKey.Trim('='));
-            }
-            else
-            {
-                //  https://github.com/google/google-authenticator/wiki/Conflicting-Accounts
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:63]: Please don't remove these comments

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 12]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
+            var validChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.~";
 
-            foreach (char symbol in value)
+            foreach (var symbol in value)
             {
-                if (validChars.IndexOf(symbol) != -1)
-                {
-                    result.Append(symbol);
-                }
+                if (validChars.IndexOf(symbol) == -1)
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:139]: Please don't remove {} here

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 13]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
+                    result.AppendFormat("%{0:X2}", (int) symbol);
                 else
-                {
-                    result.Append('%' + String.Format("{0:X2}", (int)symbol));
-                }
+                    result.Append(symbol);
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:142]: Please don't remove {} here

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 14]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
-        {
-            byte[] key = Encoding.UTF8.GetBytes(secret);
-            return GenerateHashedCode(key, iterationNumber, digits);
-        }
+        private string GenerateHashedCode(string secret,
+            long iterationNumber,
+            bool secretIsBase32,
+            int digits = 6)
+            => GenerateHashedCode(
+                secretIsBase32 ? Base32Encoding.ToBytes(secret):Encoding.UTF8.GetBytes(secret),
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:160]: 👍

File: Google.Authenticator/TwoFactorAuthenticator.cs [hunk 15]
```diff
@@ -17,209 +16,222 @@ namespace Google.Authenticator
...
-        {
-            var codes = GetCurrentPINs(accountSecretKey, timeTolerance);
-            return codes.Any(c => c == twoFactorCodeFromClient);
-        }
+        public bool ValidateTwoFactorPIN(
+            string accountSecretKey,
+            string twoFactorCodeFromClient,
+            TimeSpan timeTolerance,
+            bool secretIsBase32 = false) =>
+            GetCurrentPINs(accountSecretKey, timeTolerance, secretIsBase32).Any(c => c == twoFactorCodeFromClient);
```
Comment (flytzen) [inline Google.Authenticator/TwoFactorAuthenticator.cs:204]: I agree this is more compact, but I think it hurts legibility. I am not too bothered, but I prefer the previous form here.

**Review by flytzen — CHANGES_REQUESTED** [review]
Thank you for this and for fixing the issue.
I would prefer to leave {} around the body of if/else; it is more verbose but is recommended practice because it prevents subtle errors.
If at all possible, could you add a unit test that would have failed before the fix and passes after? Reason is, the code *looks* good, but I don't actually know if it works and a unit test would show that. 

Thank you again, great work!

File: Google.Authenticator/Base32Encoding.cs (hunk 1 above)
Comment (caterina-novak) [reply Google.Authenticator/Base32Encoding.cs:19]: Done :)

File: Google.Authenticator/Base32Encoding.cs (hunk 2 above)
Comment (caterina-novak) [reply Google.Authenticator/Base32Encoding.cs:51]: Done :)

File: Google.Authenticator/Base32Encoding.cs (hunk 3 above)
Comment (caterina-novak) [reply Google.Authenticator/Base32Encoding.cs:65]: Done

File: Google.Authenticator/Base32Encoding.cs (hunk 4 above)
Comment (caterina-novak) [reply Google.Authenticator/Base32Encoding.cs:105]: Done

File: Google.Authenticator.Tests/AuthCodeTest.cs [hunk 16]
```diff
@@ -21,5 +22,22 @@ public void BasicAuthCodeTest()
 
             actual.ShouldBe(expected);
         }
+
```
Comment (flytzen) [inline Google.Authenticator.Tests/AuthCodeTest.cs:25]: 💯

**Review by flytzen — APPROVED** [review]

File: Google.Authenticator/TwoFactorAuthenticator.cs (hunk 11 above)
Comment (caterina-novak) [reply Google.Authenticator/TwoFactorAuthenticator.cs:63]: Sorry, that was not on purpose. Returned it.

**Comment by flytzen** [comment]
@ahwm LGTM, you happy to merge?
Given that we are changing the public interface, we should probably increment the major version?
And then we need to raise @BrandonPotter to get a deploy running :)

**Comment by caterina-novak** [comment]
> 
> 
> @ahwm LGTM, you happy to merge?
> Given that we are changing the public interface, we should probably increment the major version?
> And then we need to raise @BrandonPotter to get a deploy running :)

It will be great if the Nuget package is upgraded! Then I will be able to use it in our product ;) Thank you in advance!

**Comment by BrandonPotter** [comment]
Good stuff here!

**Comment by ahwm** [comment]
> @ahwm LGTM, you happy to merge?
> Given that we are changing the public interface, we should probably increment the major version?
> And then we need to raise @BrandonPotter to get a deploy running :)

I'm happy to merge. Yes the public API is being changed, but it's been done in a backwards-compatible way (parameters with a default value last in the list) so it's not going to break existing integrations that update to it. So I think we can get by with a minor version increment. 

So instead of 2.1.2 it would be 2.2.0.

**Comment by flytzen** [comment]
@ahwm agreed, thank you 💯

**Comment by BrandonPotter** [comment]
Agreed!

**Comment by caterina-novak** [comment]
@awhm could you please update the Nuget package to include these changes? Thank you very much in advance.

**Comment by ahwm** [comment]
@caterina-novak pending #67

**Comment by ahwm** [comment]
@caterina-novak NuGet is pushed and should be live in a short period.

**Comment by caterina-novak** [comment]
@ahwm Adam, thank you very much!

Best regards,Caterina mailto: ***@***.***  

    On Tuesday, 13 July 2021, 04:36:18 EEST, Adam ***@***.***> wrote:  
 
 


@caterina-novak NuGet is pushed and should be live in a short period.

**Pull request merged**